### PR TITLE
add variables model in object_key_value storage

### PR DESCRIPTION
**Pull request by ShatalovYaroslav**

### CODE REVIEW

File: src/main/java/org/ow2/proactive/catalog/util/parser/WorkflowParser.java [hunk 1]
```diff
@@ -237,16 +242,26 @@ private void handleDescriptionElement(ImmutableList.Builder<KeyValueLabelMetadat
...
+                                        ATTRIBUTE_VARIABLE_MODEL_LABEL,
+                                        ATTRIBUTE_VARIABLE_NAME,
+                                        ATTRIBUTE_VARIABLE_MODEL,
+                                        xmlStreamReader,
+                                        false);
+
     }
 
     private void handleElementWithMultipleValues(ImmutableList.Builder<KeyValueLabelMetadataEntity> keyValueMapBuilder,
             String attributeLabel, String attributeNameForKey, String attributeNameForValue,
```
Comment (tobwiens) [inline src/main/java/org/ow2/proactive/catalog/util/parser/WorkflowParser.java]: If you can: This method needs refactoring.

File: src/main/java/org/ow2/proactive/catalog/util/parser/WorkflowParser.java [hunk 2]
```diff
@@ -258,7 +273,11 @@ private void handleElementWithMultipleValues(ImmutableList.Builder<KeyValueLabel
             }
         }, xmlStreamReader);
 
-        keyValueMapBuilder.add(new KeyValueLabelMetadataEntity(key[0], value[0], attributeLabel));
+        if (key[0] != null && value[0] != null) {
+            if (allowEmptyValues || (!value[0].isEmpty() && !key[0].isEmpty())) {
```
Comment (sonarqube-activeeon) [inline src/main/java/org/ow2/proactive/catalog/util/parser/WorkflowParser.java]: MAJOR Merge this if statement with the enclosing one. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1066)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: src/main/java/org/ow2/proactive/catalog/util/parser/WorkflowParser.java [hunk 3]
```diff
@@ -258,7 +273,11 @@ private void handleElementWithMultipleValues(ImmutableList.Builder<KeyValueLabel
             }
         }, xmlStreamReader);
 
-        keyValueMapBuilder.add(new KeyValueLabelMetadataEntity(key[0], value[0], attributeLabel));
+        if (key[0] != null && value[0] != null) {
```
Comment (jrochas) [inline src/main/java/org/ow2/proactive/catalog/util/parser/WorkflowParser.java]: I guess you could create utility functions to express better what all these conditions mean.

File: src/main/java/org/ow2/proactive/catalog/util/parser/WorkflowParser.java [hunk 4]
```diff
@@ -237,37 +244,61 @@ private void handleDescriptionElement(ImmutableList.Builder<KeyValueLabelMetadat
...
-            } else if (attributeName.equals(attributeNameForValue)) {
-                value[0] = attributeValue;
+            XMLStreamReader xmlStreamReader, boolean allowEmptyValues) {
+        List<String> strings = orderedSearch(xmlStreamReader, attributeNameForKey, attributeNameForValue);
+
+        String key = strings.get(0);
+        String value = strings.get(1);
+
+        if (checkIfNotNull(key, value)) {
+            if (allowEmptyValues || checkIfNotEmpty(key, value)) {
```
Comment (Mykhailenko) [inline src/main/java/org/ow2/proactive/catalog/util/parser/WorkflowParser.java]: Please collapse two 'if's in single 'if'.

Comment (sonarqube-activeeon) [inline src/main/java/org/ow2/proactive/catalog/util/parser/WorkflowParser.java]: MAJOR Merge this if statement with the enclosing one. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1066)

File: src/main/java/org/ow2/proactive/catalog/util/parser/WorkflowParser.java [hunk 5]
```diff
@@ -237,37 +244,61 @@ private void handleDescriptionElement(ImmutableList.Builder<KeyValueLabelMetadat
...
+                keyValueMapBuilder.add(new KeyValueLabelMetadataEntity(key, value, attributeLabel));
             }
-        }, xmlStreamReader);
+        }
+    }
+
+    private boolean checkIfNotNull(String... values) {
+        return Arrays.stream(values).noneMatch(s -> s == null);
```
Comment (sonarqube-activeeon) [inline src/main/java/org/ow2/proactive/catalog/util/parser/WorkflowParser.java]: MINOR Replace this lambda with method reference 'Objects::isNull'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1612)

File: src/main/java/org/ow2/proactive/catalog/util/parser/WorkflowParser.java [hunk 6]
```diff
@@ -237,37 +244,61 @@ private void handleDescriptionElement(ImmutableList.Builder<KeyValueLabelMetadat
...
+
+    }
+
+    private boolean checkIfNotEmpty(String... values) {
+        return Arrays.stream(values).noneMatch(s -> s.isEmpty());
```
Comment (sonarqube-activeeon) [inline src/main/java/org/ow2/proactive/catalog/util/parser/WorkflowParser.java]: MINOR Replace this lambda with a method reference. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1612)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/org/ow2/proactive/catalog/util/parser/WorkflowParser.java (hunk 5 above)
Comment (Mykhailenko) [reply src/main/java/org/ow2/proactive/catalog/util/parser/WorkflowParser.java]: Agree

File: src/main/java/org/ow2/proactive/catalog/util/parser/WorkflowParser.java (hunk 6 above)
Comment (Mykhailenko) [reply src/main/java/org/ow2/proactive/catalog/util/parser/WorkflowParser.java]: Agree

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**